core/common/src/main/java/alluxio/metrics/MetricsSystem.java

```diff
@@ -778,8 +778,17 @@ private static synchronized List<alluxio.grpc.Metric> reportMetrics(InstanceType
               entry.getKey(), gauge.getValue().getClass().getSimpleName());
           continue;
         }
-        rpcMetrics.add(Metric.from(entry.getKey(),
-            ((Number) gauge.getValue()).longValue(), MetricType.GAUGE).toProto());
+        // As the value is a number, it can also be reported to master as "COUNTER" type.
+        // Then it can also be aggregated in the same way as "COUNTER" at master side.
+        long value = ((Number) gauge.getValue()).longValue();
+        Long prev = LAST_REPORTED_METRICS.replace(entry.getKey(), value);
```

Comment:
If a worker crashes, and restarts afterwards, ` LAST_REPORTED_METRICS` is empty so the first update will not be a diff but the absolute value. This will cause the cluster aggregated value to increase over time as old values of the crashed workers are never subtracted.

Reply:
Yeah. I think the existing aggregation mechanism for `Counter` also has this issue, right?
Is there any idea/plan to fix this? Thank you.

Reply:
Just a rough idea, not sure if it will work.

How about making the cluster aggregated metric a meter, and make the workers report the absolute number of active threads in each heartbeat. The meter is supposed to tell you the average between the last 1, 5 and 15 minutes. Any obsolete data older than 15 min will be purged. So if a worker crashes, its data will linger no longer than 15min in the metric system.

Reply:
If using a meter, may I know how to sum the active threads from different workers to get a cluster view? Different clusters may have different heartbeat intervals.

Reply:
Maybe meter was not a good idea. A meter outputs a moving average of a value over the past X minutes and sth like "num of threads per second" just does not make sense.

What we need is a way to phase out the old readings from a worker that's no longer active. This can be done by using a map that tracks the worker's reported readings and remove them after a period of time for which the worker has been inactive.

The server side metric can be a gauge that reports the aggregated thread numbers weighted by the age of the report. Something like:

```java
Report {
  int reading;
  long timestamp;
  long reportingInterval;
}
Map<Worker, Report> reports;
Gauge clusterActiveReaderThreads = new Gauge(() -> {
  long weightedNumThreads = 0;
  for (worker, report) in reports {
    long age = report.timestamp - now;
    if (age > THRESHOLD) {
      reports.remove(worker);
      continue;
    }
    if (age > report.reportingInterval) {
      weightedNumThreads += report.reading * (1 - (age - interval) / (THRESHOLD - interval));
    } else {
      weightedNumThreads += report.reading;
    }  
  }
  return weightedNumThreads;
});
```

Reply:
As all metrics aggregating at master side have this issue (worker restarting reports a wrong diff value)) , I'd to create a new issue to track it. We can discuss the solution there. It is ok with you? @dbw9580

Reply:
sure go ahead!